packages/protocol/contracts/common/FeeBurner.sol

```diff
@@ -0,0 +1,425 @@
+    */
+  function burnAllCelo() public {
+    ICeloToken celo = ICeloToken(registry.getAddressForOrDie(GOLD_TOKEN_REGISTRY_ID));
+    celo.burn(celo.balanceOf(address(this)));
+  }
+
+  /**
+    * @param token the address of the token to query.
+    * @return The amount burned for a token.
+    */
+  function getPastBurnForToken(address token) external view returns (uint256) {
+    return pastBurn[token];
+  }
+
+  /**
+    * @param token the address of the token to query.
+    * @param amountToBurn the amount of the token to burn.
+    * @return Returns true if burning amountToBurn would exceed the dayli mimit.
+    */
+  function limitHit(address token, uint256 amountToBurn) public returns (bool) {
+    if (dailyBurnLimit[token] == 0) {
+      // if no limit set, assume uncapped
+      return false;
+    }
+
+    uint256 currentDay = now / 1 days;
+    // Pattern borrowed from Reserve.sol
+    if (currentDay > lastLimitDay) {
+      lastLimitDay = currentDay;
+      currentDayLimit[token] = dailyBurnLimit[token];
+    }
+
+    return amountToBurn >= currentDayLimit[token];
+  }
+
+  /**
+    * @notice Updates the current day limit for a token.
+    * @param token the address of the token to query.
+    * @param amountBurned the amount of the token that was burned.
+    */
+  function updateLimits(address token, uint256 amountBurned) private {
+    if (dailyBurnLimit[token] == 0) {
+      // if no limit set, assume uncapped
+      return;
+    }
+    currentDayLimit[token] = currentDayLimit[token].sub(amountBurned);
+    emit DailyLimitUpdatedDeleteMe(amountBurned);
+    return;
+  }
+
+  /**
+    * @notice Burns the max possible of a Mento token.
+    * @dev Should be used in case the loop fails because one token is reverting or OOG.
+    * @param tokenAddress the address of the token to burn.
+    */
+  function burnSingleMentoToken(address tokenAddress) public onlyWhenNotFrozen {
+    StableToken stableToken = StableToken(tokenAddress);
+    uint256 balanceToBurn = stableToken.balanceOf(address(this));
+
+    if (limitHit(tokenAddress, balanceToBurn)) {
+      // in case the limit is hit, burn the max possible
+      balanceToBurn = currentDayLimit[tokenAddress];
+      emit DailyLimitHit(tokenAddress, balanceToBurn);
+    }
+
+    address exchangeAddress = registry.getAddressForOrDie(stableToken.getExchangeRegistryId());
+
+    IExchange exchange = IExchange(exchangeAddress);
+
+    uint256 minAmount = 0;
+    if (FixidityLib.unwrap(maxSlippage[tokenAddress]) != 0) {
+      // max slippage is set
+      // use sorted oracles as reference
+      ISortedOracles sortedOracles = getSortedOracles();
+      (uint256 priceWithoutSlippage, ) = sortedOracles.medianRate(tokenAddress);
+      minAmount = calculateMinAmount(priceWithoutSlippage, tokenAddress, balanceToBurn);
+    }
+
+    // small numbers cause rounding errors and zero case should be skipped
+    if (balanceToBurn <= MIN_BURN) {
```

Comment:
We could probably move this condition higher - somewhere to line 244 to save some gas